src/base64.cr

```diff
@@ -244,65 +244,89 @@ module Base64
     end
   end
 
-  private def from_base64(data)
+  # Processes the given data and yields each byte.
+  private def from_base64(data : Bytes, &block : UInt8 -> Nil)
     size = data.size
-    dt = DECODE_TABLE.to_unsafe
-    cstr = data.to_unsafe
-    start_cstr = cstr
-    while (size > 0) && (sym = cstr[size - 1]) && (sym == NL || sym == NR || sym == PAD)
+    bytes = data.to_unsafe
+    bytes_begin = bytes
+
+    # Get the position of the last valid base64 character (rstrip '\n', '\r' and '=')
+    while (size > 0) && (sym = bytes[size - 1]) && (sym == NL || sym == NR || sym == PAD)
       size -= 1
     end
-    endcstr = cstr + size - 4
 
+    # Process combinations of four characters until there aren't any left
+    fin = bytes + size - 4
     while true
-      break if cstr > endcstr
-      while cstr.value == NL || cstr.value == NR
-        cstr += 1
-      end
+      break if bytes > fin
 
-      break if cstr > endcstr
-      a, b, c, d = next_decoded_value, next_decoded_value, next_decoded_value, next_decoded_value
+      # Move the pointer by one byte until there is a valid base64 character
+      while bytes.value == NL || bytes.value == NR
+        bytes += 1
+      end
+      break if bytes > fin
 
-      yield (a << 2 | b >> 4).to_u8!
-      yield (b << 4 | c >> 2).to_u8!
-      yield (c << 6 | d).to_u8!
+      yield_decoded_chunk_bytes(
+        raise_error: true,
+        chunk_pos: bytes - bytes_begin,
+        bytes: {bytes[0], bytes[1], bytes[2], bytes[3]}
+      )
+      bytes += 4
     end
 
-    while (cstr < endcstr + 4) && (cstr.value == NL || cstr.value == NR)
-      cstr += 1
+    # Move the pointer by one byte until there is a valid base64 character or the end of `bytes` was reached
+    while (bytes < fin + 4) && (bytes.value == NL || bytes.value == NR)
+      bytes += 1
     end
 
-    mod = (endcstr - cstr) % 4
-    if mod == 2
-      a, b = next_decoded_value, next_decoded_value
-      yield (a << 2 | b >> 4).to_u8!
-    elsif mod == 3
-      a, b, c = next_decoded_value, next_decoded_value, next_decoded_value
-      yield (a << 2 | b >> 4).to_u8!
-      yield (b << 4 | c >> 2).to_u8!
-    elsif mod != 0
-      raise Error.new("Wrong size")
+    # If the amount of base64 characters is not divisable by 4, the remainder of the previous loop is handled here
+    unread_bytes = (fin - bytes) % 4
```

Comment:
Why is there a modulo operation? How can `unread_bytes` be `> 3`?

Reply:
https://github.com/crystal-lang/crystal/blob/3b6e62c649f7097dec4b1e8785a6c4ea0f19726f/src/base64.cr#L275

Because it's been that way before.
I think it cannot be greater than 3, but I would have to double-check that.

Reply:
Just tested it, using `fin - bytes + 4` would also work.
the results of `fin - bytes` are in the range `-4..-1`.
I don't know if it's worth changing it.

Reply:
Ah, right. The difference is negative. I suppose the modulo operation is just used to make it positive, then.